### ISSUE TITLE
Merge Song and Sample into a single class

### ISSUE THREAD

**Issue by jlnr** [gosu/gosu#117]
See discussion here https://github.com/jlnr/gosu/issues/116

How to name it? Sound? SampleInstance could be renamed according to http://www.libgosu.org/cgi-bin/mwf/topic_show.pl?tid=624

**Comment by Kjarrigan** [comment]
I'll look into this. This is be one of the oldest issues but resolving this might resolve other issues (#390) as well or at least make things a little easier here and there ( #147 , #264 ).

**Comment by jlnr** [comment]
Hmm. I don't actually know how this issue would interact with #390. It's bad enough that `Gosu::Window` has to keep `Song.current_song` running, but how would it work if there could be any number of streaming Sounds at the same time? (Basically, Sample would become `Sound.new(filename)` and Song would be `Sound.new(filename, streaming: true)`)

**Comment by jlnr** [comment]
Also, there is actually one supported platform where `Gosu::Song` is magic: iOS. It uses a completely different API for the current song because that way it's hardware-accelerated. But if the merger works well on "normal" platforms, I'm sure I can find a way to make it work on iOS as well :)

**Comment by jlnr** [comment]
A quick braindump of suggestions:
- Sample is renamed into Sound and gains a `streaming:` constructor parameter. It then dynamically uses one of severeal implementations: the old Sample code, and whatever code Song uses when streaming a file. (Song already has a tiny class hierarchy for its implementations, with bizarro class names for historical reasons.)
- Song (for backward compatibility) is a simple wrapper around a Sound and a SampleInstance (should be renamed to Channel while we are at it - #314). When you call `Song#play`, it stores the Channel so that it can implement `Song#stop` etc.
- All *streaming* sounds add themselves to a global array while they are playing, and Song::update() (C++ only) moves to Sound::update() and lets *all* of these Sounds stream more data. Gosu::Window doesn't even need to be changed.

The alternative is to work with threads to fix #390, but I think that would too much new code in addition to all the refactoring action, and should be a separate branch.

**Comment by Kjarrigan** [comment]
> Song (for backward compatibility) is a simple wrapper around a Sound and a SampleInstance (should be renamed to Channel while we are at it - #314). When you call Song#play, it stores the Channel so that it can implement Song#stop etc.

Should this wrapper be available in C++ and Ruby?

> The alternative is to work with threads

I definitely won't work with threads.

>  It's bad enough that Gosu::Window has to keep Song.current_song running, but how would it work if there could be any number of streaming Sounds at the same time?

I wonder - was there a specific (design) reason, that only one song could be a played at a time? I mean it would probably sound really crazy if multiple songs overlap each other but would provide some interessting sound options, like playing a "base melody" the whole game and depending on the current scene just add a little flavor on top.

**Comment by jlnr** [comment]
> Should this wrapper be available in C++ and Ruby?

Good question. If we are going to deprecate Song, we might as well write the wrapper twice - as a public gist for C++ (similar to how [Gosu's sockets ended up in their own repo](https://github.com/jlnr/Sockets)) and as a bit of pure Ruby in `compat.rb`. I think it would be a nice illustration of how Sound and Channel work for new users of Gosu.

> I wonder - was there a specific (design) reason, that only one song could be a played at a time?

Yep - both FMOD 3 and SDL_mixer only supported one "song" (streaming or MID/S3M/IT/...) at the same time. Gosu used FMOD 3 for a long, long time.

**Comment by jlnr** [comment]
Can we please use Sound instead of Audio? It is more in line with other libraries (gdx and sfml, not sure about others). Gosu used to have an Audio class, but it was a counterpart to the Graphics class - a global manager for everything audible. That's also how gdx uses the names.

**Comment by Kjarrigan** [comment]
Yeah I already renamed to to Sound and Channel as you suggested but didn't push my latest "progress" yet.

**Comment by Kjarrigan** [comment]
> Gosu used to have an Audio class, but it was a counterpart to the Graphics class - a global manager for everything audible.

Why was it split into Song and Sample(Instance)? By choice or forced by the underlying audiolibs, because that's somewhat what will be the endresult of this approch - a "universal" Sound-class.

**Comment by jlnr** [comment]
No, Audio was a part of Window like Graphics or Input. It managed audio initialisation. You had to have an Audio instance to create Samples and Songs.

**Comment by Kjarrigan** [comment]
*Blow the dust away* It's been a while since I looked into it but after the latest success in the graphic department I feel confident to finally solve this issue as well. The Channel-renaming was already done. Now lets see if the Song + Sample = Sound works as intended.

**Comment by dfl** [comment]
I'd like to add some custom sample-based effects (filter, delay, reverb, etc) into my audio chain. It would be awesome if there was a function that I could override to inject my effects into the audio stream (from ruby). @Kjarrigan seems like something not too complicated to expose during the audio rewrite?

**Comment by Kjarrigan** [comment]
Well honestly I don't know. Unfortunately I haven't worked on this in quite some time and I'm not sure if this could easily be included. I haven't done any effect-audio stuff done so far, so do you have some examples how it would look like in e.g. other languages / libs?

I assume you would want something like this?
```ruby
audio = Sample.new 'cool.ogg'
audio.play do |io|
   apply_filters(io)
end
```

Internally we do something like this:
```cpp
        stream_to_buffer(buffers[0]);
        stream_to_buffer(buffers[1]);        
        alSourceQueueBuffers(source, 2, buffers);
        alSourcePlay(source);
```
Not sure if this can be extended to add room for your desired freedom.

**Comment by dfl** [comment]
What I need is the buffers to be exposed for transformation before streaming.

```
class Effect
  def initialize( samplerate )
  	...
  end
  def clear!
  	...
  end

  def tick(sample)
     ...
    return sample
  end
end

wave = Sample.new 'cool.ogg'
process = Effect.new(wave.samplerate)
wave.buffers.size.times do |i| # all channels
  process.clear
  wave.buffers[i].map!{|s| process.tick(s) }
}
wave.play
```

Ideally I could also insert my own buffers, to synthesize sounds from scratch.

```
samplerate = 44100
seconds = 1
freq = 440.0
wave = []
(samplerate*seconds).times do |t|
  wave[t] = Math.sin(2*Math::PI*freq*t)
end
sin_440 = Sample.new( AudioBuffer.new(wave, samplerate) )
sin_440.play
```

**Comment by jlnr** [comment]
Creating Samples from a buffer is realistic and consistent with the `Image.from_blob` method (see also #198). Applying effects while streaming is out of scope for Gosu, and it sounds like something that would be limited by Ruby's performance even on current machines.
